kimchi/src/snarky/traits.rs

```diff
@@ -0,0 +1,136 @@
+//! Traits.
+
+use ark_ff::PrimeField;
+
+use super::{
+    checked_runner::{RunState, WitnessGeneration},
+    cvar::CVar,
+    prelude::TypeCreation,
+};
+
+/// A snarky type is a type that can be used in a circuit.
+/// It references an equivalent "out-of-circuit" type that one can use outside of the circuit.
+/// (For example, to construct private or public inputs, or a public output, to the circuit.)
+pub trait SnarkyType<F>: Sized
+where
+    F: PrimeField,
+{
+    /// Some additional information. (TODO: add more documentation.)
```

Comment:
> Some 'out-of-circuit' data, which is carried as part of `Self`.
> This data isn't encoded as CVars in the circuit, since the data may be large (e.g. a sparse merkle tree), or may only be used by witness computations / for debugging.

Reply:
thanks!

Reply:
I still don't get how "custom" data can be compatible with the API. For example, `constraint_system_auxiliary` seems to produce the data on demand.